scrapy/core/http2/protocol.py

```diff
@@ -202,27 +198,29 @@ def _handle_events(self, events):
                 logger.debug('Received unhandled event {}'.format(event))
 
     # Event handler functions starts here
-    def data_received(self, event: DataReceived):
+    def data_received(self, event: DataReceived) -> None:
         self.streams[event.stream_id].receive_data(event.data, event.flow_controlled_length)
 
-    def response_received(self, event: ResponseReceived):
+    def response_received(self, event: ResponseReceived) -> None:
         self.streams[event.stream_id].receive_headers(event.headers)
 
-    def settings_acknowledged(self, event: SettingsAcknowledged):
+    def settings_acknowledged(self, event: SettingsAcknowledged) -> None:
         # Send off all the pending requests as now we have
         # established a proper HTTP/2 connection
         self._send_pending_requests()
 
         # Update certificate when our HTTP/2 connection is established
-        self._metadata['certificate'] = Certificate(self.transport.getPeerCertificate())
+        self.metadata['certificate'] = Certificate(self.transport.getPeerCertificate())
 
-    def stream_ended(self, event: StreamEnded):
-        self.streams[event.stream_id].close(StreamCloseReason.ENDED)
+    def stream_ended(self, event: StreamEnded) -> None:
+        stream = self.pop_stream(event.stream_id)
+        stream.close(StreamCloseReason.ENDED, from_protocol=True)
 
-    def stream_reset(self, event: StreamReset):
-        self.streams[event.stream_id].close(StreamCloseReason.RESET)
+    def stream_reset(self, event: StreamReset) -> None:
+        stream = self.pop_stream(event.stream_id)
+        stream.close(StreamCloseReason.RESET, from_protocol=True)
```

Comment:
Here the `stream_ended()` and `stream_reset()` methods should pop the stream out of the `self.streams` dict right? 
I think we can change the definition of `Stream.close(reason, error, from_protocol)` to `Stream.close(reason, error)` and pop the stream from the dict only when the reason is one of `(CONNECTION_LOST, INACTIVE)` ?